torch/_export/passes/add_runtime_assertions_for_constraints_pass.py

```diff
@@ -0,0 +1,191 @@
+
+        # Convert simple sympy Integers into concrete int to
+        # insert into graph
+        def _convert_to_int(val):
+            if val == sympy.oo:
+                return math.inf
+            if isinstance(val, sympy.Integer):
+                return int(val)
+            raise RuntimeError(
+                "Export constraints cannot be non-integer expressions"
+            )
+
+        constraint_id_to_dim: Dict[int, Dict[int, List[Tuple[int, int]]]] = defaultdict(
+            lambda: defaultdict(list)
+        )
+        for constraint in constraints:
+            min_int_val = _convert_to_int(constraint["min"])
+            max_int_val = _convert_to_int(constraint["max"])
+            constraint_id_to_dim[constraint["t_id"]][constraint["dim"]].append(
+                (min_int_val, max_int_val)
+            )
+
+        # Merge the constraints into a single list of constraints
+        for t_id in constraint_id_to_dim:
+            for dim, constraints in constraint_id_to_dim[t_id].items():
+                min_vals = [x[0] for x in constraints]
+                max_vals = [x[1] for x in constraints]
+                min_val = sorted(min_vals, reverse=True)[0]
+                max_val = sorted(max_vals, reverse=False)[0]
+
+                assert min_val <= max_val
+
+                constraints_id_to_constraint[t_id].append(
+                    ConstraintSpec(constraint_dim=dim, min_val=min_val, max_val=max_val)
+                )
+
+        return constraints_id_to_constraint
+
+    def call(self, graph_module: torch.fx.GraphModule) -> None:
+        # resets the counter
+        self.input_tracker = 0
+        self.current_gm = graph_module
+        self.constraints = self._process_constraints(get_export_meta(self.current_gm).input_shape_constraints)
+        self.example_inputs = pytree.tree_flatten(get_export_meta(self.current_gm).example_inputs)[0]
+        return super().call(graph_module)
+
+    def _insert_specialized_shapes_assert(self, arg, dims, current_inp):
+        # we don't want to get shapes from meta as they will be symbolic
+        shapes = current_inp.shape
+        for dim in dims:
+            assert_msg = (
+                f"Input #{self.input_tracker}'s dimension #{dim} size is "
+                f"specialized at {shapes[dim]}"
+            )
+            dim_node = self.call_operator(
+                torch.ops.aten.sym_size,
+                (arg, dim),
+                {},
+                NodeMetadata({}),
+            )
+            eq = self.call_operator(
+                operator.eq,
+                (dim_node, shapes[dim]),
+                {},
+                NodeMetadata({}),
+            )
+            tensor_eq = self.call_operator(
+                torch.ops.aten.scalar_tensor.default,
+                (eq,),
+                {},
+                NodeMetadata({}),
+            )
+            self.call_operator(
+                torch.ops.aten._assert_async.msg,
```

Comment:
Curious what this does: Is it inserting the assert op in the exported graph?

Reply:
it's inserting the assert op in a new graph that will be returned from running this pass.